File: core/src/passes/flag_write_elim.cpp
// core/src/passes/flag_write_elim.cpp — F1-PS-012.
//
// Drops implicit flag writes whose NZCV is never consumed.
//
// Flag writers in our IR are `CmpFlags`, `AluFlags`, `Compare`, and
// `WriteFlagsCountZero`.
// Flag READERS are `CondJumpRel` (branches on NZCV) and `Select`
// (lowers to csel, which reads the NZCV set by the most recent
// flag writer — the CMOV / BZHI / CMPXCHG / BSF decode pattern).
// The pass walks forward, tracking the most-recent flag writer; on
// each flag reader it pins the writer as needed; on a fresh flag
// writer it drops the previous one if no reader pinned it.
// End-of-stream: a still-pending droppable flag write is dropped
// (no consumer follows).
//
// Compare writes flags too, but it has a result Ref that other code
// paths consume — we never drop a Compare. CmpFlags has no result
// Ref so dropping is safe.

#include "prisma/passes.hpp"

#include <cstddef>
#include <variant>
#include <vector>

namespace prisma::passes {

std::vector<ir::Stmt>
flag_write_elimination(const std::vector<ir::Stmt>& stmts) {
    const std::size_t n = stmts.size();
    std::vector<bool> drop(n, false);

    // Index of the most recent flag writer, or n if none. We only
    // ever drop CmpFlags / AluFlags / WriteFlagsCountZero (no result
    // Ref); Compare is always kept.
    std::size_t pending_writer    = n;
    bool        pending_is_droppable = false;

    for (std::size_t i = 0; i < n; ++i) {
        const auto& st = stmts[i];
        if (std::holds_alternative<ir::CmpFlags>(st.op)
            || std::holds_alternative<ir::AluFlags>(st.op)
            || std::holds_alternative<ir::WriteFlagsCountZero>(st.op)) {
            // A new implicit flag write supersedes the previous pending writer.
            // If the previous pending writer was an unread droppable write,
            // mark it for deletion now.
            if (pending_writer != n && pending_is_droppable) {
                drop[pending_writer] = true;
            }
            pending_writer      = i;
            pending_is_droppable = true;
        } else if (std::holds_alternative<ir::Compare>(st.op)) {
            // Compare also writes flags. Any pending droppable write becomes
            // stale and droppable. Compare itself stays (its result
            // Ref may have other consumers).
            if (pending_writer != n && pending_is_droppable) {
                drop[pending_writer] = true;
            }
            pending_writer      = i;
            pending_is_droppable = false;
        } else if (std::holds_alternative<ir::CondJumpRel>(st.op)
                   || std::holds_alternative<ir::Select>(st.op)) {
            // Flag reader: pin the most recent writer. Select lowers
            // to csel and consumes NZCV exactly like a conditional
            // branch — dropping its CmpFlags left csel reading stale
            // flags on hardware (caught by the gap-sweep ARM64 e2e).
            pending_writer      = n;
            pending_is_droppable = false;
        }
    }
    // End-of-block: a still-pending droppable write has no consumer.
    if (pending_writer != n && pending_is_droppable) {
        drop[pending_writer] = true;
    }

    std::vector<ir::Stmt> out;
    out.reserve(n);
    for (std::size_t i = 0; i < n; ++i) {
        if (!drop[i]) out.push_back(stmts[i]);
    }
    return out;
}

}  // namespace prisma::passes

File: shell/prisma-passes/src/flag_write_elim.rs
//! Flag write elimination pass.
//!
//! Mirrors C++ `flag_write_elimination`. Drops implicit flag writes whose
//! NZCV is never consumed. Flag writers: `CmpFlags`, `AluFlags`, `Compare`,
//! `WriteFlagsCountZero`. Flag readers: `CondJumpRel`, `Select` (lowers to
//! csel). `Compare` is never dropped (its result ref may be consumed
//! elsewhere); only the result-less writers (`CmpFlags`, `AluFlags`,
//! `WriteFlagsCountZero`) are droppable.

use prisma_ir::{BasicBlock, Function, Op};

use crate::Pass;

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct FlagWriteElim;

impl FlagWriteElim {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Pass for FlagWriteElim {
    fn name(&self) -> &'static str {
        "flag_write_elimination"
    }

    fn run(&self, func: Function) -> Function {
        flag_write_elimination(func)
    }
}

/// Per-block flag write elimination.
pub fn flag_write_elimination(func: Function) -> Function {
    let blocks = func
        .blocks
        .into_iter()
        .map(|block| {
            let n = block.stmts.len();
            let mut drop = vec![false; n];
            // Index of the most-recent flag writer; None if none pending.
            let mut pending_writer: Option<usize> = None;
            let mut pending_droppable = false;

            for (i, stmt) in block.stmts.iter().enumerate() {
                match &stmt.op {
                    Op::CmpFlags(_) | Op::AluFlags(_) | Op::WriteFlagsCountZero(_) => {
                        if let Some(prev) = pending_writer {
                            if pending_droppable {
                                drop[prev] = true;
                            }
                        }
                        pending_writer = Some(i);
                        pending_droppable = true;
                    }
                    Op::Compare(_) => {
                        if let Some(prev) = pending_writer {
                            if pending_droppable {
                                drop[prev] = true;
                            }
                        }
                        pending_writer = Some(i);
                        pending_droppable = false;
                    }
                    Op::CondJumpRel(_) | Op::Select(_) => {
                        // Reader pins the most recent writer.
                        pending_writer = None;
                        pending_droppable = false;
                    }
                    _ => {}
                }
            }
            // End-of-block: a still-pending droppable write has no consumer.
            if let Some(prev) = pending_writer {
                if pending_droppable {
                    drop[prev] = true;
                }
            }

            let stmts = block
                .stmts
                .into_iter()
                .enumerate()
                .filter_map(|(i, s)| if drop[i] { None } else { Some(s) })
                .collect();

            BasicBlock { id: block.id, stmts }
        })
        .collect();

    Function {
        blocks,
        entry: func.entry,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use prisma_ir::{CmpFlags, CondCode, CondJumpRel, OpSize, Select, Stmt};

    fn cmp(lhs: u32, rhs: u32) -> Stmt {
        Stmt::new(None, Op::CmpFlags(CmpFlags { lhs, rhs, size: OpSize::I64 }))
    }

    fn cond_jump() -> Stmt {
        Stmt::new(
            None,
            Op::CondJumpRel(CondJumpRel {
                cc: CondCode::Eq,
                target_guest_pc: 1,
                fallthrough_guest_pc: 2,
            }),
        )
    }

    fn block(stmts: Vec<Stmt>) -> Function {
        Function { entry: 0, blocks: vec![BasicBlock { id: 0, stmts }] }
    }

    #[test]
    fn unconsumed_cmp_is_dropped() {
        let out = flag_write_elimination(block(vec![cmp(0, 1)]));
        assert!(out.blocks[0].stmts.is_empty());
    }

    #[test]
    fn consumed_cmp_is_kept() {
        let out = flag_write_elimination(block(vec![cmp(0, 1), cond_jump()]));
        assert_eq!(out.blocks[0].stmts.len(), 2);
    }

    #[test]
    fn superseded_cmp_is_dropped_but_last_kept_if_read() {
        // cmp A ; cmp B ; cond_jump  -> first cmp dropped, second kept.
        let out = flag_write_elimination(block(vec![cmp(0, 1), cmp(2, 3), cond_jump()]));
        assert_eq!(out.blocks[0].stmts.len(), 2);
        match &out.blocks[0].stmts[0].op {
            Op::CmpFlags(c) => assert_eq!((c.lhs, c.rhs), (2, 3)),
            other => panic!("unexpected {other:?}"),
        }
    }

    #[test]
    fn select_reads_flags_and_pins_writer() {
        let out = flag_write_elimination(block(vec![
            cmp(0, 1),
            Stmt::new(
                Some(9),
                Op::Select(Select { cc: CondCode::Eq, true_value: 0, false_value: 1, size: OpSize::I64 }),
            ),
        ]));
        assert_eq!(out.blocks[0].stmts.len(), 2);
    }

    #[test]
    fn compare_is_never_dropped() {
        use prisma_ir::Compare;
        let out = flag_write_elimination(block(vec![Stmt::new(
            Some(2),
            Op::Compare(Compare { cc: CondCode::Eq, lhs: 0, rhs: 1, size: OpSize::I64 }),
        )]));
        assert_eq!(out.blocks[0].stmts.len(), 1);
    }
}
